### PR TITLE
srp_daemon: Fix pkey handling

### PR DESCRIPTION
**Pull request by bvanassche**

### CODE REVIEW

File: libibverbs/verbs.c [hunk 1]
```diff
@@ -200,6 +200,22 @@ LATEST_SYMVER_FUNC(ibv_query_pkey, 1_1, "IBVERBS_1.1",
 	return 0;
 }
 
+LATEST_SYMVER_FUNC(ibv_get_pkey_index, 1_1, "IBVERBS_1.1",
```
Comment (jgunthorpe) [inline libibverbs/verbs.c]: libibverbs/libibverbs.map.in needs to be updated with this symbol, and the version should be IBVERBS_1.2
Same for debian/debian/libibverbs1.symbols

CI will fail because of this.

Comment (bvanassche) [reply libibverbs/verbs.c]: From libibverbs.map.in: /* NOTE: The next stanza for public symbols should be IBVERBS_1.4 due to release 12 */. I'm confused ...

Comment (bvanassche) [reply libibverbs/verbs.c]: I'm not familiar with rdma-core shared library versioning. Can you have a look at the update I pushed out?

Comment (jgunthorpe) [reply libibverbs/verbs.c]: Looks like the note in the map file is almost correct (woops), release 12 removed the IB_VERBS_1.3 and 1.4 stanza's.

So the next lowest unused number is IBVERBS_1.5, which is what you should use here.

File: libibverbs/libibverbs.map.in [hunk 2]
```diff
@@ -80,6 +80,7 @@ IBVERBS_1.1 {
 		ibv_get_device_guid;
 		ibv_get_device_list;
 		ibv_get_device_name;
+		ibv_get_pkey_index;
```
Comment (jgunthorpe) [inline libibverbs/libibverbs.map.in]: This needs its own stanza:

```c
IBVERBS_1.5 {
       global:
                         ibv_get_pkey_index;
} IBVERBS_1.1;
```

File: libibverbs/verbs.c (hunk 1 above)
Comment (jgunthorpe) [inline libibverbs/verbs.c]: And this should be:

```c
LATEST_SYMVER_FUNC(ibv_get_pkey_index, 1_5, "IBVERBS_1.5",
```

File: srp_daemon/srp_ib_types.h [hunk 3]
```diff
@@ -91,26 +91,28 @@ enum {
...
+struct ib_path_rec {
 	uint8_t		resv0[8];
 	union umad_gid	dgid;
 	union umad_gid	sgid;
 	__be16		dlid;
 	__be16		slid;
 	__be32		hop_flow_raw;
 	uint8_t		tclass;
-	uint8_t		num_path;
+	uint8_t		reversible_numpath; /* reversible-7:7 num path-6:0 */
```
Comment (jgunthorpe) [inline srp_daemon/srp_ib_types.h]: This hunk is unrelated?

File: debian/libibverbs1.symbols [hunk 4]
```diff
@@ -57,6 +57,7 @@ libibverbs.so.1 libibverbs1 #MINVER#
  ibv_get_device_list@IBVERBS_1.1 1.1.6
  ibv_get_device_name@IBVERBS_1.0 1.1.6
  ibv_get_device_name@IBVERBS_1.1 1.1.6
+ ibv_get_pkey_index@IBVERBS_1.1 1.2.2
```
Comment (jgunthorpe) [inline debian/libibverbs1.symbols]: Should be 

```
ibv_get_pkey_index@IBVERBS_1.5 19
````
and need a line at the top like

```
IBVERBS_1.5@IBVERBS_1.5 19
```

File: libibverbs/libibverbs.map.in (hunk 2 above)
Comment (jgunthorpe) [reply libibverbs/libibverbs.map.in]: And the CMakeLists.txt needs to be modified:

```
rdma_library(ibverbs "${CMAKE_CURRENT_BINARY_DIR}/libibverbs.map"
  # See Documentation/versioning.md
  1 1.5.${PACKAGE_VERSION}
```

Comment (bvanassche) [reply libibverbs/libibverbs.map.in]: An additional change was needed. Can you have another look? Thanks, Bart.

File: debian/libibverbs1.symbols [hunk 5]
```diff
@@ -1,7 +1,8 @@
 libibverbs.so.1 libibverbs1 #MINVER#
  IBVERBS_1.0@IBVERBS_1.0 1.1.6
  IBVERBS_1.1@IBVERBS_1.1 1.1.6
- (symver)IBVERBS_PRIVATE_19 19
+ IBVERBS_1.5@IBVERBS_1.5 19
```
Comment (jgunthorpe) [inline debian/libibverbs1.symbols]: Oh we are at version 20 now? That last 19 should be 20, sorry.

File: CMakeLists.txt [hunk 6]
```diff
@@ -60,7 +60,7 @@ set(PACKAGE_NAME "RDMA")
 set(PACKAGE_VERSION "20.0")
 # When this is changed the values in these files need changing too:
 #   debian/libibverbs1.symbols
-set(IBVERBS_PABI_VERSION "19")
+set(IBVERBS_PABI_VERSION "20")
```
Comment (jgunthorpe) [inline CMakeLists.txt]: Why? This is only changed when the IBVERBS_PRIVATE symbols are altered, I didn't see any of those change in this series?

**Comment by jgunthorpe** [comment]
When changing the map file/etc it is necessary to check travis's Debian build to make sure the shlibs file is OK, it says this:

```
dpkg-gensymbols: warning: some new symbols appeared in the symbols file: see diff output below
dpkg-gensymbols: warning: debian/libibverbs1/DEBIAN/symbols doesn't match completely debian/libibverbs1.symbols
--- debian/libibverbs1.symbols (libibverbs1_20.0-1_amd64)
+++ dpkg-gensymbolst3JeIX	2018-07-26 22:56:58.358019189 +0000
@@ -1,6 +1,7 @@
 libibverbs.so.1 libibverbs1 #MINVER#
  IBVERBS_1.0@IBVERBS_1.0 1.1.6
  IBVERBS_1.1@IBVERBS_1.1 1.1.6
+ IBVERBS_1.5@IBVERBS_1.5 20.0-1
  (symver)IBVERBS_PRIVATE_20 20
  ibv_ack_async_event@IBVERBS_1.0 1.1.6
  ibv_ack_async_event@IBVERBS_1.1 1.1.6
````

Otherwise this is OK as far as I can tell

**Comment by bvanassche** [comment]
How to address that warning? I'm not familiar with shared library symbol versioning.

**Comment by jgunthorpe** [comment]
Add the line it suggests to  debian/libibverbs1.symbols

```
IBVERBS_1.5@IBVERBS_1.5 20.
```

But without the Debian minor versions

**Comment by jgunthorpe** [comment]
Updated with a comment change to libibverbs.map
